### PR TITLE
Annotate more CodeAnalysis public APIs

### PR DESCRIPTION
**Pull request by jcouv**
Picked more yet-to-be-annotated APIs from https://github.com/dotnet/roslyn/pull/41000

### CODE REVIEW

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs [hunk 1]
```diff
@@ -143,13 +144,6 @@ private CompilationWithAnalyzers(Compilation compilation, ImmutableArray<Diagnos
             _executingCompilationOrNonConcurrentTreeTask = null;
         }
 
-        private void AddExceptionDiagnostic(Exception exception, DiagnosticAnalyzer analyzer, Diagnostic diagnostic)
```
Comment (jcouv) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs]: >AddExceptionDiagnostic [](start = 21, length = 22)

:memo: removed unused method (same below)

File: src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs [hunk 2]
```diff
@@ -55,8 +58,13 @@ private sealed class SequenceEqualComparer : IEqualityComparer<List<Section>>
         {
             public static SequenceEqualComparer Instance { get; } = new SequenceEqualComparer();
 
-            public bool Equals(List<Section> x, List<Section> y)
+            public bool Equals([AllowNull] List<Section> x, [AllowNull] List<Section> y)
             {
+                if (x is null || y is null)
+                {
+                    return x is null && y is null;
+                }
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs]: Can simplify to 

```
if (x is null)
{
  return y is null;
}
```

Comment (jcouv) [reply src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs]: I don't think that's tight enough. If only `y` is null, then we would NRE  on the dereference on next line.

File: src/Compilers/CSharp/Portable/Symbols/Attributes/PEAttributeData.cs [hunk 3]
```diff
@@ -17,8 +19,8 @@ internal sealed class PEAttributeData : CSharpAttributeData
     {
         private readonly MetadataDecoder _decoder;
         private readonly CustomAttributeHandle _handle;
-        private NamedTypeSymbol _lazyAttributeClass = ErrorTypeSymbol.UnknownResultType; // Indicates uninitialized.
-        private MethodSymbol _lazyAttributeConstructor;
+        private NamedTypeSymbol? _lazyAttributeClass = ErrorTypeSymbol.UnknownResultType; // Indicates uninitialized.
```
[the remark this answers is not in the document]
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/Attributes/PEAttributeData.cs]: We assign `null` to `_lazyAttributeClass` in some cases in `EnsureClassAndConstructorSymbolsAreLoaded`, but never explicitly. I think that's why we're using a sentinel value, as opposed to `null` to track the uninitialized state.

File: src/Compilers/CSharp/Portable/Symbols/Attributes/AttributeData.cs [hunk 4]
```diff
@@ -118,8 +122,8 @@ internal bool IsSecurityAttribute(CSharpCompilation compilation)
                 // Well-known type SecurityAttribute is optional.
                 // Native compiler doesn't generate a use-site error if it is not found, we do the same.
                 var wellKnownType = compilation.GetWellKnownType(WellKnownType.System_Security_Permissions_SecurityAttribute);
-                HashSet<DiagnosticInfo> useSiteDiagnostics = null;
-                _lazyIsSecurityAttribute = AttributeClass.IsDerivedFrom(wellKnownType, TypeCompareKind.ConsiderEverything, useSiteDiagnostics: ref useSiteDiagnostics).ToThreeState();
+                HashSet<DiagnosticInfo>? useSiteDiagnostics = null;
+                _lazyIsSecurityAttribute = AttributeClass!.IsDerivedFrom(wellKnownType, TypeCompareKind.ConsiderEverything, useSiteDiagnostics: ref useSiteDiagnostics).ToThreeState();
```
[the remark this answers is not in the document]
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/Attributes/AttributeData.cs]: I think you're right, this should be an assertion instead.

File: src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs (hunk 2 above)
Comment (jaredpar) [reply src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs]: Doh that's right.

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for a second review, please.

**Pull request merged**